src/Analyser/MutatingScope.php

```diff
@@ -3466,23 +3470,38 @@ public function specifyExpressionType(Expr $expr, Type $type, Type $nativeType):
 					if ($dimType instanceof ConstantIntegerType) {
 						$types[] = new StringType();
 					}
+
+					// keep certainty
+					$variableCertainty = null;
+					$hasExpressionType = $scope->hasExpressionType($expr->var);
+					if (!$hasExpressionType->no()) {
+						$variableCertainty = $hasExpressionType;
```

Comment:
This looks a bit fishy. Why is the parameter handled different than for other expression types? I don't like that for ArrayDimFetch, you're basically throwing away what the user passed in when calling specifyExpressionType.